### ISSUE TITLE
XSLT: debugging template rule selection

### ISSUE THREAD

**Issue by michaelhkay** [qt4cg/qtspecs#1153]
The biggest headache when debugging XSLT stylesheets is working out which template rules have been invoked in response to an `xsl:apply-templates` instruction (I'm hitting this frustration right now with the qtspecs build stylesheets...). The `xsl:message` instruction is unhelpful here, because if the "wrong" template rule is firing, you don't know where to add the message. And the only other standardised debugging aids are `fn:trace()` and `xsl:assert`, which don't help either.

I propose an attribute on `xsl:apply-templates`, `xsl:apply-imports`, and `xsl:next-match`: trace=yes|no. If enabled, execution of the instruction causes a message to be output (as if by xsl:message) identifying the rule that is invoked, in an implementation-defined way. In the case that a built-in template rule is invoked, the message should indicate this, and any implicit apply-templates performed by the built-in rule should be evaluated as if it specified trace="yes". It is "recommended" that the message output should identify the stylesheet module, line number, match pattern, and mode, if the information is available, and should also include a representation of the item that is being processed by the instruction, for example the node kind and name.

**Comment by dnovatchev** [comment]
Not doubting the value of such a feature, we need to give it a more precise name - such as "**_logging_**".
My experience shows that logging can produce huge output quickly, and can become unwieldly in many cases.

True debugging involves more interactivity rather than the "job submission and execution" style associated with logging.

Short of a true XSLT and XPath IDE, one could dream of an **`<xsl:breakpoint>`** instruction, and why not something like  **#breakpoint#** in XPath?

The execution processor must have a way to be run in "interactive mode" and to initiate a dialog upon hitting a breakpoint. This dialog could allow the user to issue a few queries that return value(s) of variables, stacks (of initiated template rules or of function calls or of nested `let/return` expressions whose evaluation is in progress, and to add dynamically new breakpoints or remove/disable currently active ones without the need to start the transformation/evaluation anew from the very beginning (or at least not to leave such perception in the end-user/developer).

**Comment by michaelhkay** [comment]
Yes, interactive debugging of this kind of thing can be very useful, but I think it's outside our scope. In many cases (like the one I was dealing with yesterday) it's a challenge first to isolate a specific transformation from a complex workflow, which you need to do before interactive debugging can start. I did actually solve this one with logging (saxon:trace="yes" on the `xsl:mode` element, which produced about a million lines of output; a trace at the level of the `apply-templates` instruction would have been much more manageable).

My own debugging experience is very atypical because most of the time, if I'm debugging something then it's because a fault in the XSLT processor is suspected. So I'm interested to know what "ordinary users" do, and what features would make their life easier.

**Comment by ndw** [comment]
In my experience (personally and watching other users and developers work on large stylesheets like the DocBook ones), users typically add `xsl:message` statements more and more aggressively until they locate the troublesome template.

In that respect, I think a trace facility would be very helpful. What's less clear is exactly how it should be scoped. I think tracing should inherit so that I can set it on `xsl:stylesheet` or `xsl:mode` and have it apply to all the contained templates (except those where I've explicitly set it to `no`, of course). I guess it follows that it should also be inherted "down" into imported and included modules. So I can set it on the very tipy-top root stylesheet element to see *everything* or on an individual module or mode to see only those templates.

**Comment by Arithmeticus** [comment]
I like the idea. 

Just to explore our options, what about instead we introduce `xsl:with-message` as a permitted child of  `xsl:apply-templates`, `xsl:apply-imports`, and `xsl:next-match`, on analogy with `xsl:with-param`? In addition to the standard attributes allowed for `xsl:message`, it would have the following:

Attribute `@expires`, default value 1, but any positive integer, indicating the maximum number of times depthwise the message should be fired as the mode is traversed.  If zero or less, the message is fired throughout the lifespan of template traversal (same longevity as a tunnel parameter).

Attribute `@prepend-trace-info` (or something like that) `yes|no` default `yes` would specify that the message should be prepended by info about the stylesheet module, line number, match pattern, mode, etc. 

Special record variable `$trace-info` -- a map of the relevant trace info. This would allow someone to customize the body of their messages.

One bonus about an element-based approach, over an attribute one, is that if you've set up multiple traces, and they fire in the same downstream template, the body of the message will help you disambiguate which one has been fired.

**Comment by dnovatchev** [comment]
One way to deal with the huge volume of logging is to have a standard output format - why not in XML format? - which then can be queried as per the developer needs.

So, I run the transformation and the logged output is in one or more XML files. Then I run my XPath/XSLT/ XQuery - maybe from within an IDE and narrow down to only the wanted information.

If the XML format is well-designed, this could be input to a "pseudo-debugger" that would use it to simulate an actual evaluation.

Just thinking out loud ...

**Comment by dnovatchev** [comment]
Another useful feature would be similar to the "data breakpoints" implemented more than 10 years ago in Marrowsoft XSelerator and (I think) also in Microsoft's XSLT 1.0 Debugger.

This would produce logging output whenever a particular node of a given XML document is accessed - not only when a template is applied on this node, but whenever the node is referenced within an XPath expression during its evaluation.

**Comment by ndw** [comment]
It's possible to imagine more and more elaborate tools, but for the specification, it might be best to focus on "the minimum needed to declare victory."

**Comment by dnovatchev** [comment]
> It's possible to imagine more and more elaborate tools, but for the specification, it might be best to focus on "the minimum needed to declare victory."

True.

At the same time, the idea to produce logging in a standard XML format is not so wild and "more and more elaborate" - it is just taking care that the invisible-XML people will have less to do in the future - at least regarding the XSLT / XQuery / XPath logging content.

It would also give the reader a way to further transform / filter /combine the content of any such message in order to get whatever they are interested in in their particular situation - which cannot be anticipated and imagined in advance.

If the logging is just a step in the problem-solving processing pipeline, making this as easy and flexible for further processing is really a no-brainer.

**Comment by michaelhkay** [comment]
Let's leave this one for implementors to add extensions. Propose closing with no action.
